lib/util.h

```diff
@@ -0,0 +1,57 @@
+#ifndef _LIB_UTIL_H_
+#define _LIB_UTIL_H_ 1
+
+// `nelems (a)` is meant to give the number of elements in `a` and should only be used when `a` is an array.
+//
+// Usage:
+//
+//   ```
+//   char buffer [512];
+//   snprintf (buffer, nelems (buffer), "The values are (%f, %f).", x, y);
+//   ```
+//
+//   This is much safer than the version using `sizeof (buffer)`
+//   (which creates a bad habit -- invitation to overflow
+//   when switching to versions of `snprintf` for `wchar_t` etc.)
+//   and is more compact than the version using `sizeof (buffer) / sizeof (buffer [0])`.
+//
+//   Also, unlike the `snprintf (buffer, 512, etc.)` version, it avoids repetition of `512`.
+//
+//   Also, unlike the following version:
+//   ```
+//   const size_t buffer_size = 512;
+//   char buffer [buffer_size];
+//   snprintf (buffer, buffer_size, "The values are (%f, %f).", x, y);
+//   ```
+//   it avoids introducing a new identifier.
+//
+// Rationale for the current implementation:
+// Using `sizeof ((a) [0])` allows `a` to be a variable with user-defined `operator[]`.
+// Using `sizeof (0 [a])` rules out user-defined `operator[]` and requires `a` to be an array or, unfortunately, a pointer.
+// Thanks to Matthew Wilson (of STLsoft fame) for this.
+//
+// TODO: We should only allow `a` to be an array (i.e. we should rule out pointers).
+
+#ifndef nelems
+    #define nelems(a) (sizeof (a) / sizeof (0 [a]))
```

Comment:
we have a definition of this macro in algorithms.h too.

Reply:
> we have a definition of this macro in algorithms.h too.

Thank you. I am seeing it now:

`#define ELEMENTS(a)   (sizeof(a) / sizeof(a[0]))`

I think it should be:

`#define ELEMENTS(a)   (sizeof(a) / sizeof((a)[0]))`

or better (in order to prevent `a` from being an object with user-defined `operator[]`, which would not give what the user probably expects):

`#define ELEMENTS(a)   (sizeof(a) / sizeof(0[a]))`

Reply:
Note an even better implementation (for C++11 and above) is just:
template<typename t, size_t i>
constexpr inline size_t sizeofarray(const t (&arr)[i])  { return i; }

No reasons for macros.

Reply:
Note for C++17+, std::size works too.
See https://en.cppreference.com/w/cpp/iterator/size .

Reply:
> Note an even better implementation (for C++11 and above) is just: template<typename t, size_t i> constexpr inline size_t sizeofarray(const t (&arr)[i]) { return i; }
> 
> No reasons for macros.

Thank you.

I believe your suggestion is also safer, preventing errors such as invoking the macro not with (a reference to) an array, but instead with (a reference to) a pointer (possibly after array-to-pointer conversion, as for argument-to-parameter of called function).